// payload/dist/fields/config/sanitize.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sanitize_1 = __importDefault(require("./sanitize"));
const errors_1 = require("../../errors");
describe('sanitizeFields', () => {
    it('should throw on missing type field', () => {
        const fields = [{
                label: 'some-collection',
                name: 'Some Collection',
            }];
        expect(() => {
            sanitize_1.default(fields, []);
        }).toThrow(errors_1.MissingFieldType);
    });
    describe('relationships', () => {
        it('should not throw on valid relationship', () => {
            const validRelationships = ['some-collection'];
            const fields = [{
                    type: 'relationship',
                    label: 'my-relationship',
                    name: 'My Relationship',
                    relationTo: 'some-collection',
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).not.toThrow();
        });
        it('should not throw on valid relationship - multiple', () => {
            const validRelationships = ['some-collection', 'another-collection'];
            const fields = [{
                    type: 'relationship',
                    label: 'my-relationship',
                    name: 'My Relationship',
                    relationTo: ['some-collection', 'another-collection'],
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).not.toThrow();
        });
        it('should not throw on valid relationship inside blocks', () => {
            const validRelationships = ['some-collection'];
            const fields = [{
                    name: 'layout',
                    label: 'Layout Blocks',
                    labels: {
                        singular: 'Block',
                    },
                    type: 'blocks',
                    blocks: [{
                            fields: [{
                                    type: 'relationship',
                                    label: 'my-relationship',
                                    name: 'My Relationship',
                                    relationTo: 'some-collection',
                                }],
                        }],
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).not.toThrow();
        });
        it('should throw on invalid relationship', () => {
            const validRelationships = ['some-collection'];
            const fields = [{
                    type: 'relationship',
                    label: 'my-relationship',
                    name: 'My Relationship',
                    relationTo: 'not-valid',
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).toThrow(errors_1.InvalidFieldRelationship);
        });
        it('should throw on invalid relationship - multiple', () => {
            const validRelationships = ['some-collection', 'another-collection'];
            const fields = [{
                    type: 'relationship',
                    label: 'my-relationship',
                    name: 'My Relationship',
                    relationTo: ['some-collection', 'not-valid'],
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).toThrow(errors_1.InvalidFieldRelationship);
        });
        it('should throw on invalid relationship inside blocks', () => {
            const validRelationships = ['some-collection'];
            const fields = [{
                    name: 'layout',
                    label: 'Layout Blocks',
                    labels: {
                        singular: 'Block',
                    },
                    type: 'blocks',
                    blocks: [{
                            fields: [{
                                    type: 'relationship',
                                    label: 'my-relationship',
                                    name: 'My Relationship',
                                    relationTo: 'not-valid',
                                }],
                        }],
                }];
            expect(() => {
                sanitize_1.default(fields, validRelationships);
            }).toThrow(errors_1.InvalidFieldRelationship);
        });
    });
});
