// payload/dist/config/defaults.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaults = void 0;
const path_1 = __importDefault(require("path"));
exports.defaults = {
    defaultDepth: 2,
    maxDepth: 10,
    collections: [],
    globals: [],
    cookiePrefix: 'payload',
    csrf: [],
    cors: [],
    admin: {
        meta: {
            titleSuffix: '- Payload',
        },
        disable: false,
        indexHTML: path_1.default.resolve(__dirname, '../admin/index.html'),
        components: {},
        css: path_1.default.resolve(__dirname, '../admin/scss/custom.css'),
        scss: path_1.default.resolve(__dirname, '../admin/scss/overrides.scss'),
    },
    upload: {},
    graphQL: {
        mutations: {},
        queries: {},
        maxComplexity: 1000,
        disablePlaygroundInProduction: true,
    },
    routes: {
        admin: '/admin',
        api: '/api',
        graphQL: '/graphql',
        graphQLPlayground: '/graphql-playground',
    },
    rateLimit: {
        window: 15 * 60 * 100,
        max: 500,
    },
    express: {
        json: {},
        compression: {},
        middleware: [],
    },
    hooks: {},
    localization: false,
};
