// payload/dist/config/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const schema_1 = __importDefault(require("../collections/config/schema"));
const schema_2 = __importDefault(require("../globals/config/schema"));
const component = joi_1.default.alternatives().try(joi_1.default.object().unknown(), joi_1.default.func());
exports.default = joi_1.default.object({
    serverURL: joi_1.default.string()
        .required(),
    cookiePrefix: joi_1.default.string(),
    routes: joi_1.default.object({
        admin: joi_1.default.string(),
        api: joi_1.default.string(),
        graphQL: joi_1.default.string(),
        graphQLPlayground: joi_1.default.string(),
    }),
    collections: joi_1.default.array()
        .items(schema_1.default),
    globals: joi_1.default.array()
        .items(schema_2.default),
    admin: joi_1.default.object({
        user: joi_1.default.string(),
        meta: joi_1.default.object()
            .keys({
            titleSuffix: joi_1.default.string(),
            ogImage: joi_1.default.string(),
            favicon: joi_1.default.string(),
        }),
        disable: joi_1.default.bool(),
        indexHTML: joi_1.default.string(),
        css: joi_1.default.string(),
        scss: joi_1.default.string(),
        components: joi_1.default.object()
            .keys({
            Nav: component,
            Dashboard: component,
            graphics: joi_1.default.object({
                Icon: component,
                Logo: component,
            }),
        }),
        webpack: joi_1.default.func(),
    }),
    defaultDepth: joi_1.default.number()
        .min(0)
        .max(30),
    maxDepth: joi_1.default.number()
        .min(0)
        .max(100),
    csrf: joi_1.default.array()
        .items(joi_1.default.string()),
    cors: [
        joi_1.default.string()
            .valid('*'),
        joi_1.default.array()
            .items(joi_1.default.string()),
    ],
    publicENV: joi_1.default.object(),
    express: joi_1.default.object()
        .keys({
        json: joi_1.default.object(),
        compression: joi_1.default.object(),
        middleware: joi_1.default.array().items(joi_1.default.object()),
    }),
    local: joi_1.default.boolean(),
    upload: joi_1.default.object()
        .keys({
        limits: joi_1.default.object()
            .keys({
            fileSize: joi_1.default.number(),
        }),
    }),
    serverModules: joi_1.default.array()
        .items(joi_1.default.string()),
    rateLimit: joi_1.default.object()
        .keys({
        window: joi_1.default.number(),
        max: joi_1.default.number(),
        trustProxy: joi_1.default.boolean(),
        skip: joi_1.default.func(),
    }),
    graphQL: joi_1.default.object()
        .keys({
        mutations: joi_1.default.object(),
        queries: joi_1.default.object(),
        maxComplexity: joi_1.default.number(),
        disablePlaygroundInProduction: joi_1.default.boolean(),
        disable: joi_1.default.boolean(),
    }),
    localization: joi_1.default.alternatives()
        .try(joi_1.default.object().keys({
        locales: joi_1.default.array().items(joi_1.default.string()),
        defaultLocale: joi_1.default.string(),
        fallback: joi_1.default.boolean(),
    }), joi_1.default.boolean()),
    email: joi_1.default.alternatives()
        .try(joi_1.default.object({
        transport: 'mock',
        fromName: joi_1.default.string(),
        fromAddress: joi_1.default.string(),
    }), joi_1.default.object({
        transport: joi_1.default.object(),
        transportOptions: joi_1.default.object(),
        fromName: joi_1.default.string(),
        fromAddress: joi_1.default.string(),
    })),
    hooks: joi_1.default.object().keys({
        afterError: joi_1.default.func(),
    }),
});
