// payload/dist/fields/baseFields/baseAPIKeyFields.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = __importDefault(require("crypto"));
const encryptKey = ({ req, value }) => (value ? req.payload.encrypt(value) : undefined);
const decryptKey = ({ req, value }) => (value ? req.payload.decrypt(value) : undefined);
exports.default = [
    {
        name: 'enableAPIKey',
        type: 'checkbox',
        defaultValue: false,
        admin: {
            disabled: true,
        },
    },
    {
        name: 'apiKey',
        type: 'text',
        admin: {
            disabled: true,
        },
        hooks: {
            beforeChange: [
                encryptKey,
            ],
            afterRead: [
                decryptKey,
            ],
        },
    },
    {
        name: 'apiKeyIndex',
        type: 'text',
        hidden: true,
        admin: {
            disabled: true,
        },
        hooks: {
            beforeValidate: [
                async ({ data, req, value }) => {
                    if (data.apiKey) {
                        return crypto_1.default.createHmac('sha1', req.payload.secret)
                            .update(data.apiKey)
                            .digest('hex');
                    }
                    if (data.enableAPIKey === false) {
                        return null;
                    }
                    return value;
                },
            ],
        },
    },
];
