// payload/dist/mongoose/buildSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable no-use-before-define */
const mongoose_1 = require("mongoose");
const errors_1 = require("../errors");
const setBlockDiscriminators = (fields, schema) => {
    fields.forEach((field) => {
        const blockFieldType = field;
        if (blockFieldType.type === 'blocks' && blockFieldType.blocks && blockFieldType.blocks.length > 0) {
            blockFieldType.blocks.forEach((blockItem) => {
                let blockSchemaFields = {};
                blockItem.fields.forEach((blockField) => {
                    const fieldSchema = fieldToSchemaMap[blockField.type];
                    if (fieldSchema) {
                        blockSchemaFields = fieldSchema(blockField, blockSchemaFields);
                    }
                });
                const blockSchema = new mongoose_1.Schema(blockSchemaFields, { _id: false });
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore Possible incorrect typing in mongoose types, this works
                schema.path(field.name).discriminator(blockItem.slug, blockSchema);
                setBlockDiscriminators(blockItem.fields, blockSchema);
            });
        }
    });
};
const formatBaseSchema = (field) => {
    const createAccess = field.access && field.access.create;
    const condition = field.admin && field.admin.condition;
    return {
        localized: field.localized || false,
        unique: field.unique || false,
        required: (field.required && !field.localized && !condition && !createAccess) || false,
        default: field.defaultValue || undefined,
        index: field.index || field.unique || false,
    };
};
const buildSchema = (configFields, options = {}) => {
    let fields = {};
    configFields.forEach((field) => {
        const fieldSchema = fieldToSchemaMap[field.type];
        if (fieldSchema) {
            fields = fieldSchema(field, fields);
        }
    });
    const schema = new mongoose_1.Schema(fields, options);
    setBlockDiscriminators(configFields, schema);
    return schema;
};
const fieldToSchemaMap = {
    number: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: Number },
    }),
    text: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: String },
    }),
    email: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: String },
    }),
    textarea: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: String },
    }),
    richText: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: mongoose_1.Schema.Types.Mixed },
    }),
    code: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: String },
    }),
    radio: (field, fields) => {
        if (!field.options || field.options.length === 0) {
            throw new errors_1.MissingFieldInputOptions(field);
        }
        const schema = {
            ...formatBaseSchema(field),
            type: String,
            enum: field.options.map((option) => {
                if (typeof option === 'object')
                    return option.value;
                return option;
            }),
        };
        return {
            ...fields,
            [field.name]: schema,
        };
    },
    checkbox: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: Boolean },
    }),
    date: (field, fields) => ({
        ...fields,
        [field.name]: { ...formatBaseSchema(field), type: Date },
    }),
    upload: (field, fields) => ({
        ...fields,
        [field.name]: {
            ...formatBaseSchema(field),
            type: mongoose_1.Schema.Types.ObjectId,
            ref: field.relationTo,
        },
    }),
    relationship: (field, fields) => {
        let schema = {};
        if (Array.isArray(field.relationTo)) {
            schema._id = false;
            schema.value = {
                type: mongoose_1.Schema.Types.ObjectId,
                refPath: `${field.name}${field.localized ? '.{{LOCALE}}' : ''}.relationTo`,
            };
            schema.relationTo = { type: String, enum: field.relationTo };
        }
        else {
            schema = {
                ...formatBaseSchema(field),
            };
            schema.type = mongoose_1.Schema.Types.ObjectId;
            schema.ref = field.relationTo;
        }
        if (field.hasMany) {
            schema = {
                type: [schema],
                localized: field.localized || false,
            };
        }
        return {
            ...fields,
            [field.name]: schema,
        };
    },
    row: (field, fields) => {
        const newFields = { ...fields };
        field.fields.forEach((rowField) => {
            const fieldSchemaMap = fieldToSchemaMap[rowField.type];
            if (fieldSchemaMap) {
                const fieldSchema = fieldSchemaMap(rowField, fields);
                newFields[rowField.name] = fieldSchema[rowField.name];
            }
        });
        return newFields;
    },
    array: (field, fields) => {
        const schema = buildSchema(field.fields, { _id: false, id: false });
        return {
            ...fields,
            [field.name]: {
                ...formatBaseSchema(field),
                type: [schema],
            },
        };
    },
    group: (field, fields) => {
        const schema = buildSchema(field.fields, { _id: false, id: false });
        return {
            ...fields,
            [field.name]: {
                ...formatBaseSchema(field),
                required: field.fields.some((subField) => subField.required === true),
                type: schema,
            },
        };
    },
    select: (field, fields) => {
        if (!field.options || field.options.length === 0) {
            throw new errors_1.MissingFieldInputOptions(field);
        }
        const schema = {
            ...formatBaseSchema(field),
            type: String,
            enum: field.options.map((option) => {
                if (typeof option === 'object')
                    return option.value;
                return option;
            }),
        };
        return {
            ...fields,
            [field.name]: field.hasMany ? [schema] : schema,
        };
    },
    blocks: (field, fields) => {
        const flexibleSchema = new mongoose_1.Schema({ blockName: String }, { discriminatorKey: 'blockType', _id: false, id: false });
        return {
            ...fields,
            [field.name]: {
                type: [flexibleSchema],
                localized: field.localized || false,
            },
        };
    },
};
exports.default = buildSchema;
