// payload/dist/mongoose/connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable no-console */
const mongoose_1 = __importDefault(require("mongoose"));
const logger_1 = __importDefault(require("../utilities/logger"));
const logger = logger_1.default();
const connectMongoose = async (url) => {
    let urlToConnect = url;
    let successfulConnectionMessage = 'Connected to Mongo server successfully!';
    if (process.env.NODE_ENV === 'test') {
        // eslint-disable-next-line global-require
        const { MongoMemoryServer } = require('mongodb-memory-server');
        const mongod = new MongoMemoryServer({
            instance: {
                port: 27018,
                dbName: 'payloadmemory',
            },
        });
        urlToConnect = await mongod.getUri();
        successfulConnectionMessage = 'Connected to in-memory Mongo server successfully!';
    }
    try {
        await mongoose_1.default.connect(urlToConnect, {
            useNewUrlParser: true,
            useUnifiedTopology: true,
            useCreateIndex: true,
            autoIndex: false,
        });
        logger.info(successfulConnectionMessage);
    }
    catch (err) {
        logger.error('Error: cannot connect to MongoDB. Details: ', err);
        process.exit(1);
    }
};
exports.default = connectMongoose;
