// payload/dist/fields/traverseFields.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validationPromise_1 = __importDefault(require("./validationPromise"));
const accessPromise_1 = __importDefault(require("./accessPromise"));
const hookPromise_1 = __importDefault(require("./hookPromise"));
const types_1 = require("./config/types");
const traverseFields = (args) => {
    const { fields, data = {}, originalDoc = {}, path, reduceLocales, locale, fallbackLocale, accessPromises, operation, overrideAccess, req, id, relationshipPopulations, depth, currentDepth, hook, hookPromises, fullOriginalDoc, fullData, validationPromises, errors, payload, showHiddenFields, } = args;
    fields.forEach((field) => {
        const dataCopy = data;
        if (operation === 'read' && field.hidden && typeof data[field.name] !== 'undefined' && !showHiddenFields) {
            delete data[field.name];
        }
        if ((field.type === 'upload' || field.type === 'relationship')
            && (data[field.name] === '' || data[field.name] === 'none' || data[field.name] === 'null')) {
            dataCopy[field.name] = null;
        }
        if (field.type === 'checkbox') {
            if (data[field.name] === 'true')
                dataCopy[field.name] = true;
            if (data[field.name] === 'false')
                dataCopy[field.name] = false;
            if (data[field.name] === '')
                dataCopy[field.name] = false;
        }
        if (field.type === 'richText' && typeof data[field.name] === 'string') {
            dataCopy[field.name] = JSON.parse(data[field.name]);
        }
        const hasLocalizedValue = (typeof data[field.name] === 'object' && data[field.name] !== null)
            && field.name
            && field.localized
            && locale !== 'all'
            && reduceLocales;
        if (hasLocalizedValue) {
            let localizedValue = data[field.name][locale];
            if (typeof localizedValue === 'undefined' && fallbackLocale)
                localizedValue = data[field.name][fallbackLocale];
            if (typeof localizedValue === 'undefined')
                localizedValue = null;
            dataCopy[field.name] = localizedValue;
        }
        accessPromises.push(accessPromise_1.default({
            data,
            originalDoc,
            field,
            operation,
            overrideAccess,
            req,
            id,
            relationshipPopulations,
            depth,
            currentDepth,
            hook,
            payload,
        }));
        hookPromises.push(hookPromise_1.default({
            data,
            field,
            hook,
            req,
            operation,
            fullOriginalDoc,
            fullData,
        }));
        if (types_1.fieldHasSubFields(field)) {
            if (field.name === undefined) {
                traverseFields({
                    ...args,
                    fields: field.fields,
                });
            }
            else if (types_1.fieldIsArrayType(field)) {
                if (Array.isArray(data[field.name])) {
                    data[field.name].forEach((rowData, i) => {
                        const originalDocRow = originalDoc && originalDoc[field.name] && originalDoc[field.name][i];
                        traverseFields({
                            ...args,
                            fields: field.fields,
                            data: rowData,
                            originalDoc: originalDocRow || undefined,
                            path: `${path}${field.name}.${i}.`,
                        });
                    });
                }
            }
            else {
                traverseFields({
                    ...args,
                    fields: field.fields,
                    data: data[field.name],
                    originalDoc: originalDoc[field.name],
                    path: `${path}${field.name}.`,
                });
            }
        }
        if (types_1.fieldIsBlockType(field)) {
            if (Array.isArray(data[field.name])) {
                data[field.name].forEach((rowData, i) => {
                    const block = field.blocks.find((blockType) => blockType.slug === rowData.blockType);
                    const originalDocRow = originalDoc && originalDoc[field.name] && originalDoc[field.name][i];
                    if (block) {
                        traverseFields({
                            ...args,
                            fields: block.fields,
                            data: rowData,
                            originalDoc: originalDocRow || undefined,
                            path: `${path}${field.name}.${i}.`,
                        });
                    }
                });
            }
        }
        if ((operation === 'create' || operation === 'update') && field.name) {
            const updatedData = data;
            if (data[field.name] === undefined && originalDoc[field.name] === undefined && field.defaultValue) {
                updatedData[field.name] = field.defaultValue;
            }
            if (field.type === 'array' || field.type === 'blocks') {
                const hasRowsOfNewData = Array.isArray(data[field.name]);
                const newRowCount = hasRowsOfNewData ? data[field.name].length : 0;
                // Handle cases of arrays being intentionally set to 0
                if (data[field.name] === '0' || data[field.name] === 0 || data[field.name] === null) {
                    updatedData[field.name] = [];
                }
                const hasRowsOfExistingData = Array.isArray(originalDoc[field.name]);
                const existingRowCount = hasRowsOfExistingData ? originalDoc[field.name].length : 0;
                validationPromises.push(() => validationPromise_1.default({
                    errors,
                    hook,
                    newData: { [field.name]: newRowCount },
                    existingData: { [field.name]: existingRowCount },
                    field,
                    path,
                }));
            }
            else {
                validationPromises.push(() => validationPromise_1.default({
                    errors,
                    hook,
                    newData: data,
                    existingData: originalDoc,
                    field,
                    path,
                }));
            }
        }
    });
};
exports.default = traverseFields;
